File: app/src/main/java/com/siavash/dualcamera/activities/PhotoActivity.java
package com.siavash.dualcamera.activities;

import android.app.FragmentManager;
import android.os.Bundle;
import android.support.design.widget.NavigationView;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.widget.TextView;

import com.siavash.dualcamera.Constants;
import com.siavash.dualcamera.R;
import com.siavash.dualcamera.fragments.OnFragmentInteractionListener;
import com.siavash.dualcamera.fragments.PhotoFragment;
import com.siavash.dualcamera.fragments.ShareFragment;
import com.siavash.dualcamera.util.FragmentUtil;

import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by sia on 10/21/15.
 */
public class PhotoActivity extends AppCompatActivity implements NavigationView.OnNavigationItemSelectedListener, OnFragmentInteractionListener {
    private static final String TAG = MainActivity.class.getSimpleName();

    @Bind(R.id.drawer_layout) DrawerLayout drawer;
    @Bind(R.id.toolbar) Toolbar toolbar;
    @Bind(R.id.toolbar_title) TextView toolbarTitle;
    @Bind(R.id.nav_view) NavigationView navigationView;

    private PhotoFragment photoFragment;

    @Override protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_photo);
        ButterKnife.bind(this);
        setSupportActionBar(toolbar);

        photoFragment = PhotoFragment.getInstance();
        switchFragmentTo(Constants.PHOTO_FRAGMENT);

        navigationView.setNavigationItemSelectedListener(this);
    }

    @OnClick(R.id.toolbar_hamburger) void onToolbarToggleClick() {
        if (drawer.isDrawerOpen(GravityCompat.END)) {
            drawer.closeDrawer(GravityCompat.END);
        } else {
            drawer.openDrawer(GravityCompat.END);
        }
    }

    @Override public void switchFragmentTo(int fragmentName, String... optionalValues) {
        FragmentManager fragmentManager = getFragmentManager();
        if (fragmentName == Constants.PHOTO_FRAGMENT) {
            FragmentUtil.replaceFragment(fragmentManager, R.id.container, photoFragment);
        } else if (fragmentName == Constants.SHARE_FRAGMENT) {
            FragmentUtil.replaceFragment(fragmentManager, R.id.container, ShareFragment.newInstance(optionalValues[0]));
        }
    }

    @Override public boolean onNavigationItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.nav_photo_fragment) {
            switchFragmentTo(Constants.PHOTO_FRAGMENT);
        } else if (id == R.id.nav_share_fragment) {
            switchFragmentTo(Constants.SHARE_FRAGMENT);
        }

        drawer.closeDrawer(GravityCompat.END);
        return true;
    }

    public void setToolbarTitle(String title) {
        toolbarTitle.setText(title);
    }
}

File: app/src/main/kotlin/com/siavash/dualcamera/activities/PhotoActivity.kt
package com.siavash.dualcamera.activities

import android.content.Intent
import android.os.Bundle
import android.support.design.widget.NavigationView
import android.support.v4.view.GravityCompat
import android.support.v4.widget.DrawerLayout
import android.support.v7.app.AppCompatActivity
import android.view.MenuItem
import android.view.SurfaceView
import com.siavash.dualcamera.R
import com.siavash.dualcamera.fragments.OnFragmentInteractionListener
import com.siavash.dualcamera.fragments.PhotoFragment
import com.siavash.dualcamera.fragments.ShareFragment
import com.siavash.dualcamera.util.*
import org.jetbrains.anko.intentFor

/**
 * Created by sia on 10/31/15.
 */
class PhotoActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener, OnFragmentInteractionListener {
    val drawer: DrawerLayout by bindView(R.id.drawer_layout)
    val toolbar: Toolbar by bindView(R.id.toolbar)
    val navigationView: NavigationView by bindView(R.id.nav_view)

    var photoFragment: PhotoFragment? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_photo)
        setSupportActionBar(toolbar)

        switchFragmentTo(FragmentId.PHOTO)
        navigationView.setNavigationItemSelectedListener(this)
        toolbar.defaultHamburgerAction(drawer)
        toolbar.defaultTitleStyle()
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.nav_camera_fragment) {
            startActivity(intentFor<MainActivity>().addFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT))
            finish()
        } else if (id == R.id.nav_photo_fragment) {
            switchFragmentTo(FragmentId.PHOTO)
        } else if (id == R.id.nav_share_fragment) {
            switchFragmentTo(FragmentId.SHARE)
        }
        drawer.closeDrawer(GravityCompat.END)
        return true
    }

    override fun switchFragmentTo(fragmentId: FragmentId, vararg optionalValues: String) {
        if (fragmentId.name == getCurrentFragmentTag()) return

        val fragment = fragmentManager.findFragmentByTag(fragmentId.name)
        if (fragmentId == FragmentId.PHOTO) {
            if (fragment is PhotoFragment)
                fragmentManager.popBackStack()
            else {
                photoFragment = PhotoFragment()
                fragmentManager.addFragment(R.id.container, photoFragment as PhotoFragment, fragmentId.name)
            }
        } else if (fragmentId == FragmentId.SHARE) {
            photoFragment?.saveBitmapHidden(finalImageUrl)
            fragmentManager.addFragment(R.id.container, ShareFragment(), FragmentId.SHARE.name)
        }
    }

    private fun getCurrentFragmentTag(): String? {
        if (fragmentManager.backStackEntryCount > 0) {
            return fragmentManager.getBackStackEntryAt(fragmentManager.backStackEntryCount - 1).name
        } else
            return ""
    }
}